modules/andes-core/broker/src/main/java/org/wso2/andes/thrift/MBThriftServer.java

```diff
@@ -92,12 +101,49 @@ public void start(final String hostName,
      * Stop the server
      */
     public void stop() {
+
+        log.info("Stopping the Message Broker Thrift server...");
+
         if (server != null) {
             server.stop();
+
+            try {
+                boolean terminationSuccessful = awaitTermination(THRIFT_SERVER_REQUEST_TIMEOUT, TimeUnit.SECONDS);
+
+                if (!terminationSuccessful) {
+                    log.warn("Could not stop the Message Broker Thrift server");
+                }
+            } catch (InterruptedException e) {
+                log.warn("Thrift server shutdown was interrupted", e);
+                Thread.currentThread().interrupt();
+            }
         }
     }
 
     /**
+     * Blocks until server stop serving after a stop request, or the timeout occurs, or the current thread is
+     * interrupted, whichever happens first.
+     *
+     * @param timeout the maximum time to wait
+     * @param unit the time unit of the timeout argument
+     * @return {@code true} if thrift server stopped and
+     *         {@code false} if the timeout elapsed before server stops
+     * @throws InterruptedException if interrupted while waiting
+     */
+    private boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+        final long timeoutInMillis = unit.toMillis(timeout);
+        final long shutDownStartTime = System.currentTimeMillis();
+        long timeElapsed = 0;
+
+        while (server.isServing() && timeElapsed < timeoutInMillis) {
+            TimeUnit.MILLISECONDS.sleep(100);
+            timeElapsed = System.currentTimeMillis() - shutDownStartTime;
+        }
+
+        return timeElapsed < timeoutInMillis;
```

Comment:
Isn't there a possibility where server is stopped and timeElapsed > timeoutInMillis. In that case we are returning false here and a warning will be printed. SHouldn't we return server.isServing?